common/source/CipherEngine.cpp

```diff
@@ -0,0 +1,227 @@
+    {
+        EVP_CIPHER_CTX *ctx = nullptr;
+        CipherString encoded_string_result = {};
+        CipherBytes bytes_result = {};
+        
+        int len;   
+        
+        //setting IV for the resulting obj:
+        for(int i = 0; i < 16; i++)
+        {
+            bytes_result.iv[i] = iv[i];
+        }
+ 
+
+       vector<unsigned char> msg_unsigned(msg.begin(), msg.end());
+
+        /* Create and initialise the context */
+        if(!(ctx = EVP_CIPHER_CTX_new()))
+            handleErrors();
+
+        /*
+         * Initialise the encryption operation.
+         * IMPORTANT - ensure you use a key
+         * and IV size appropriate for your cipher
+         * In this example we are using 256 bit AES
+         * (i.e. a 256 bit key). The
+         * IV size for *most* modes is the same
+         * as the block size. For AES this
+         * is 128 bits
+         */
+        if(1 != EVP_EncryptInit_ex(ctx, EVP_aes_256_cbc(), NULL, key, iv))
+            handleErrors();
+
+        /*
+         * Provide the message to be encrypted, and obtain
+         * the encrypted output.
+         * EVP_EncryptUpdate can be called multiple times if necessary
+         */
+        if(1 != EVP_EncryptUpdate(ctx, bytes_result.encrypted_msg, &len, msg_unsigned.data(), msg_unsigned.size()))
+            handleErrors();
+        bytes_result.encrypted_msg_len = len;
+
+        /*
+         * Finalise the encryption. Further ciphertext bytes 
+         * may be written at
+         * this stage.
+         */
+        if(1 != EVP_EncryptFinal_ex(ctx, bytes_result.encrypted_msg + len, &len))
+            handleErrors();
+        bytes_result.encrypted_msg_len += len;
+
+        /* Clean up */
+        EVP_CIPHER_CTX_free(ctx);
+
+
+        //Assigning values to encoded_string_result
+        encoded_string_result.encrypted_msg = encode64(bytes_result.encrypted_msg, bytes_result.encrypted_msg_len);
+        encoded_string_result.iv = encode64(bytes_result.iv, 16); 
+        encoded_string_result.encrypted_msg_len = bytes_result.encrypted_msg_len;
+    
+        return encoded_string_result;
+    }
+
+    CipherEngine::CipherString CipherEngine::encrypt(unsigned char *iv,const string& msg)
+    { 
+       CipherString result; 
+
+       result = encrypt_algorithm(iv, msg);
+
+       return result;
+    }
+
+
+    CipherEngine::CipherString CipherEngine::encrypt(const string& msg)
+    {
+        unsigned char iv[16] = {};
+        generateIV(iv);
+        
+        CipherString result;
+
+        result = encrypt_algorithm(iv, msg);
+        
+        return result;
+    }
+
+    std::string CipherEngine::decrypt(const CipherString& encoded_encrypted_string)
+    {
+    
+    EVP_CIPHER_CTX *ctx = nullptr;
+
+    //unsigned char* ciphertext;
+    int ciphertext_len = encoded_encrypted_string.encrypted_msg_len;
+    //unsigned char* iv;
+
+    int len;
```

Comment:
Move this down to where it is being used. Try to keep variables as closed to where they are being used as possible.